src/lib.rs

```diff
@@ -0,0 +1,209 @@
+const WORD_SIZE:usize = 64;
+
+#[derive(Debug)]
+pub struct PackedVec {
+    len: usize,
+    bits: Vec<u64>,
+    item_width: usize,
+}
+
+#[derive(Debug)]
+pub struct PackedVecIter<'a> {
+    packed_vec: &'a PackedVec,
+    idx: usize,
+}
+
+impl<'a> Iterator for PackedVecIter<'a> {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.idx += 1;
+        self.packed_vec.get(self.idx - 1)
+    }
+}
+
+impl <'a> PackedVec {
+    /// Return the value at the specified `index`
+    /// # Example
+    /// ```
+    /// use compressed_dst::PackedVec;
+    /// let v = vec![1, 2, 3, 4];
+    /// let small_vec = PackedVec::from_u64_vec(v);
+    /// assert_eq!(small_vec.get(3), Some(4));
+    /// ```
+    pub fn get(&self, index: usize) -> Option<u64> {
+        if index >= self.len {
+            return None;
+        }
+        let item_index = (index * self.item_width) / WORD_SIZE;
+        let start = (index * self.item_width) % WORD_SIZE;
+        if start + self.item_width < WORD_SIZE {
+            let mask = ((1 << self.item_width) - 1)
+                << (WORD_SIZE - self.item_width - start);
+            let item = (self.bits[item_index] & mask) >>
+                (WORD_SIZE - self.item_width - start);
+            Some(item)
+        } else if self.item_width == WORD_SIZE {
+            Some(self.bits[item_index])
+        } else {
+            let bits_written = WORD_SIZE - start;
+            let mask = ((1 << bits_written) - 1)
+                << (WORD_SIZE - bits_written - start);
+            let first_half = (self.bits[item_index] & mask)
+                >> (WORD_SIZE - bits_written - start);
+            // second half
+            let remaining_bits = self.item_width - bits_written;
+            if remaining_bits > 0 {
+                let mask = ((1 << remaining_bits) - 1)
+                    << (WORD_SIZE - remaining_bits);
+                let second_half = (self.bits[item_index + 1] & mask)
+                    >> (WORD_SIZE - remaining_bits);
+                Some((first_half << remaining_bits) + second_half)
+            } else {
+                Some(first_half)
+            }
+        }
+    }
+
+    /// Return a `PackedVec` containing a compressed version of the elements of
+    /// `vec`.
+    pub fn from_u64_vec(vec: Vec<u64>) -> PackedVec {
+        let len = vec.len();
+        let item_width = if let Some(v) = vec.iter().max() {
+            let log = (*v as f64).log2();
+            if log == log.ceil() {
+                (log + 1.0) as usize
+            } else {
+                log.ceil() as usize
+            }
+        } else {
+            return PackedVec {
+                len: 0,
+                bits: vec![],
+                item_width: 0,
+            };
+        };
+        let mut bit_vec = vec![];
+        let mut buff: u64 = 0;
```

Comment:
It's more common to call this just `buf`

Reply:
Good to know!